### PR TITLE
Support Solidity 0.5.x

### PR DESCRIPTION
**Pull request by area**
Based on the [leapdao](https://github.com/leapdao/solidity-coverage#master) fork, with a nod to #311 as well. Also trying to get dependencies up-to-date.

This branch is currently published under the beta tag; you can try it out by `npm install --save --dev solidity-coverage@beta`. Bug reports would be very welcome before I release properly. You must be using solidity 0.5.1 or higher (0.5.0 will not work).

### CODE REVIEW

**Comment by hiddentao** [comment]
Noticed one issue. If I have a line within a function like this:

```solidity
       address payable owner = msg.sender;
```

It seems to fail with:

```
TypeError: Cannot read property 'expressionList' of null
    at ASTBuilder.FunctionCall (/Users/ram/dev/clients/noblocknoparty/contracts/node_modules/solidity-parser-antlr/dist/ASTBuilder.js:272:17)
```

I debugged and found that it's the coverage-generated line just before that source line which fails. Specifically, it's unable to fetch function call arguments:

```solidity
emit __CoverageDeployer('/home/user/dev/contracts/Deployer.sol',31);
emit __StatementCoverageDeployer('/home/user/dev/contracts/Deployer.sol',1);  // <-- this line fails!
address payable owner = msg.sender;
```

**Comment by area** [comment]
Hm, I'm not able to reproduce it with a contract with just that line in... can you provide more context?

**Comment by hiddentao** [comment]
@area Sure.

Here is the full code:

```solidity
pragma solidity ^0.5.4;

import './zeppelin/lifecycle/Destructible.sol';
import './Conference.sol';

contract Deployer is Destructible {
    event NewParty(
        address indexed deployedAddress,
        address indexed deployer
    );

    function deploy(
        string calldata _name,
        uint256 _deposit,
        uint _limitOfParticipants,
        uint _coolingPeriod
    ) external {
       address payable owner = msg.sender;

        Conference c = new Conference(
            _name,
            _deposit,
            _limitOfParticipants,
            _coolingPeriod,
            owner
        );

        emit NewParty(address(c), owner);
    }
}
```

I'm fairly sure the code for `Destructible` and `Conference` aren't relevant to this issue, but in case you want to see them the full code is also at https://github.com/wearekickback/contracts/tree/master/contracts

**Comment by hiddentao** [comment]
Ah, just an additional note that the code above is new code that's not yet tested in the wild - I'm trying to upgrade our contracts to 0.5+, hence why I'm trying out this beta version of solidity-coverage. Thanks for the working on this, btw :)

**Comment by hiddentao** [comment]
I've re-coded it to get it working (I removed the offending line and simply used `msg.sender` directly). I had to also fix the web3 ABI decoder as a final step (see #319). It's all good now, though it would still be nice to fix this issue.

File: lib/coverageMap.js [hunk 1]
```diff
@@ -3,7 +3,7 @@
  * This file contains methods that produce a coverage map to pass to instanbul
  * from data generated by `instrumentSolidity.js`
  */
-const SolidityCoder = require('web3/lib/solidity/coder.js');
+const SolidityCoder = require('web3-eth-abi');
```
Comment (vdrg) [inline lib/coverageMap.js]: @area It looks like the API for this package has changed. I got it working as follows:

```suggestion
const { AbiCoder } = require('web3-eth-abi');
const SolidityCoder = AbiCoder();
```

Comment (vdrg) [reply lib/coverageMap.js]: Oh I guess this is related to @hiddentao's comment. I'm not sure if this is the only fix needed, but for me it is generating the reports correctly.
Also, are there plans to update testrpc-sc? There are things like `eth_signTypedData` that are present in ganache but not in the sc fork.

**Comment by area** [comment]
@hiddentao I have seen the behaviour locally, but am unable to get a test to fail along those lines, which is quite frustrating. I agree this should be fixed though, so will be holding off a non-beta release until I track it down.

File: lib/coverageMap.js (hunk 1 above)
Comment (area) [reply lib/coverageMap.js]: Hm, this change definitely breaks for me, with `TypeError: AbiCoder is not a function` - indeed AbiCoder is not defined at all. Are you working with the beta release or this branch?

I agree an update for testrpc-sc would be good. I'm going to concentrate on this solidity 0.5.x support for now though.

**Comment by area** [comment]
It looks like your error is something to do with `eth-gas-reporter`. It can't deal with an instrumented contract with that line in it, for whatever reason (even though it is valid solidity - it parses fine for coverage, obviously). Updating the version of `solidity-parser-antlr` in `eth-gas-reporter` solved the issue for me.

**Comment by area** [comment]
I think I have now successfully switched away from our fork of `solidity-parser` and moved to `solidity-parser-antlr`. I suspect there are still some patterns not correctly parsed by `parser.js`, but the tests are good now. We only had a fork to support a custom `{{XYZ}}` pattern in gnosis contracts, but they don't seem to use them any more [as far as I can see](https://github.com/gnosis/pm-contracts). I also think it's not really our responsibility to cater for that even if they do want to do that...

File: lib/coverageMap.js (hunk 1 above)
Comment (hiddentao) [reply lib/coverageMap.js]: Beta release 46 changes the API, hence why the current code breaks if you install the latest web3 abi beta. Unfortunately the web3 team's current release versioning usage breaks semantic versioning. Hence why my PR to fix the dependency version specifically.

Comment (area) [reply lib/coverageMap.js]: Oh derp, I understand now. I could have sworn your PR fixed the version of `web3`, not `web3-eth-abi`.

**Comment by area** [comment]
Hm, got an OBOE floating around in beta-1, so that release is broken!

**Comment by area** [comment]
`solidity-parser-antlr` uses `.map` [while traversing a tree](https://github.com/federicobond/solidity-parser-antlr/blob/a7bceba29d0d3d17a3379c5d6d0d7d78235cca93/src/ASTBuilder.js#L1141), so the `return false` that we had while walking the tree in the preprocessor only halted visiting the contents of that branch; if two branches had blocks that needed preprocessing, trying to do both at the same time would mess up one of them. 

Preprocessor has been altered to cope, and a test case added.

**Comment by area** [comment]
With this latest build, I get the same branch, line and function coverage as I did before on our ([Colony](https://github.com/JoinColony/colonyNetwork/). The number of things considered statements has increased very slightly; I'm not 100% sure what is responsible for that, but seems pretty innocuous (line and branch coverage are the most important metrics).

**Comment by vdrg** [comment]
Another thing that is broken is using a different compiler with truffle (using `compileCommand` in `.solcover.js`).

With truffle version ^5.0.0, now it is possible to specify the compiler version in `truffle-config.js`. When running a truffle command that compiles your contracts, truffle starts downloading the specified compiler. If the compiler was downloaded previous to running `solidity-coverage`, it works fine. However, if `solidity-coverage` is run BEFORE the compiler is downloaded, it doesn't wait for the download process to complete and crashes because it tries to compile the contracts with the default version (0.5.3).

I tested this using solc 0.5.4 in truffle, but `solidity-coverage` tried to use 0.5.3.

**Comment by area** [comment]
Hm, I think this is a `truffle.js` vs `truffle-config.js` issue. From the [documentation](https://truffleframework.com/docs/truffle/reference/configuration) at least, it seems like only the latter is supported in Truffle 5 (whereas on v4 and below, `truffle-config.js` was only expected in a windows environment).

**Comment by area** [comment]
Hm, I'm now no longer sure. Is the codebase you're seeing this with public? I'm having difficulty reproducing locally, @vdrg

**Comment by vdrg** [comment]
Not public yet, will setup a new repo to reproduce later today

File: lib/preprocessor.js [hunk 2]
```diff
@@ -42,47 +48,48 @@ module.exports.run = function r(contract) {
...
-      const ast = SolidityParser.parse(contract);
+      const ast = SolidityParser.parse(contract, { range: true });
       keepRunning = false;
-      SolExplore.traverse(ast, {
-        enter(node, parent) { // eslint-disable-line no-loop-func
-          // If consequents
-          if (node.type === 'IfStatement' && node.consequent.type !== 'BlockStatement') {
-            contract = blockWrap(contract, node.consequent);
+      blocksToWrap = [];
+      SolidityParser.visit(ast, {
```
Comment (peara) [inline lib/preprocessor.js]: Hi @area , could you also remove the while loop in this PR?
I have made a PR to do that but in the old and forked version: https://github.com/leapdao/solidity-coverage/pull/3/files
It relies on the assumption that nodes are visited in order as in the contract string.

**Comment by obernardovieira** [comment]
Hello,

I'm having `There was a problem instrumenting ./coverageEnv/contracts/Contract.sol: TypeError: Cannot read property 'stop' of null` and I don't know where it comes from yet. I'll try to find out. IN the meantime, does anyone had this problem before?

**Comment by area** [comment]
Hm, I've not seen that at any point. It looks like a parser error - can you provide the contract it's failing on? It might be a bug in `solidity-parser-antlr`.

**Comment by obernardovieira** [comment]
> Hm, I've not seen that at any point. It looks like a parser error - can you provide the contract it's failing on? It might be a bug in `solidity-parser-antlr`.

I can't, unfortunately, it's still private, but might be released soon. In the meantime, I will try to find where it failed exactly and I will get back to this issue.

**Comment by JirkaChadima** [comment]
Hi, thanks for this PR, it's much needed. However, I have a bug to report, although I'm not sure this is the right place. We have a project that's setup with solidity 0.5.6, [truffle 5.0.7](https://github.com/trufflesuite/truffle) and [zos](https://github.com/zeppelinos/zos/) to use upgradeable contracts. ZOS offers some [convenience methods for test environment](https://docs.zeppelinos.org/docs/testing.html#writing-the-test-script). And we have the test init just like they describe in their docs.

If we are running `truffle test` against `ganache-cli`, everything works as expected. However when we run `truffle test` against `testrpc-sc` with the beta release of solidity-coverage, everything breaks with

```
     Error: Provider not set or invalid
      at Object.InvalidProvider (/tmp/wt-contracts/node_modules/web3-core-helpers/src/errors.js:38:16)
      at RequestManager.send (/tmp/wt-contracts/node_modules/web3-core-requestmanager/src/index.js:128:32)
      at sendRequest (/tmp/wt-contracts/node_modules/web3-core-method/src/index.js:560:42)
      at Net.send [as getId] (/tmp/wt-contracts/node_modules/web3-core-method/src/index.js:581:13)
      at Function.<anonymous> (/tmp/wt-contracts/node_modules/zos-lib/src/artifacts/ZWeb3.ts:127:28)
      at Generator.next (<anonymous>)
      at /tmp/wt-contracts/node_modules/zos-lib/lib/artifacts/ZWeb3.js:7:71
      at new Promise (<anonymous>)
      at __awaiter (/tmp/wt-contracts/node_modules/zos-lib/lib/artifacts/ZWeb3.js:3:12)
      at Function.getNetwork (/tmp/wt-contracts/node_modules/zos-lib/lib/artifacts/ZWeb3.js:139:16)
      at Function.<anonymous> (/tmp/wt-contracts/node_modules/zos-lib/src/artifacts/ZWeb3.ts:131:35)
      at Generator.next (<anonymous>)
...
```
Now, that's indeed an error coming from zos, but I think it has something to do with something the solidity-coverage is doing with relation to code isolation. Because the provider is initially set and then just disappears from the instance.

---
Repro
```
git clone git@github.com:windingtree/wt-contracts.git
cd wt-contracts
git checkout 2a48f4c50b635ff2a74ae0910b806006f036f21d
npm install
npm run coverage
```

**Comment by area** [comment]
@JirkaChadima According to the docs for [ZOS](https://docs.zeppelinos.org/docs/2.1.0/testing.html#docsNav), you have to set `NODE_ENV=test` for things to work correctly, which you should be able to do using the `testCommand` option in `.solcover.js`.

**Comment by JirkaChadima** [comment]
> @JirkaChadima According to the docs for [ZOS](https://docs.zeppelinos.org/docs/2.1.0/testing.html#docsNav), you have to set `NODE_ENV=test` for things to work correctly, which you should be able to do using the `testCommand` option in `.solcover.js`.

Thanks for the tip, but that doesn't cut it. I've tried to run it with `.solcover.js` below, but it still spits out the same error. What is interesting is, that if I comment out copyPackages section, I get a *different* error that says that a setup transaction is reverted - which is probably tied to a buggish behaviour of zos itself.

```js
module.exports = {
  testCommand: 'NODE_ENV=test truffle test --network coverage',
  copyPackages: [
    'zos-lib'
  ]
}
```
Maybe the coverage environment is somehow picking up a different web3 version? I can setup a minimal test repo, if you're interested.

**Comment by area** [comment]
So the reason the setup transaction reverted is because ZOS are very cutting-edge, and have already introduced the `CREATE2` opcode in their assembly (which was new in Constantinople), which the modified `ganache` we were using didn't support. The only way to fix that was to bring the modified ganache up-to-date... which I have done!

I have published a new beta which should work for you, @JirkaChadima. I think you also need to add `zos` to the `copyPackages` array in `.solcover.js`, as well as the test command you have already tried. See also the note about `--allowUnlimitedContractSize`, which applies to this repository as well. Note that the increased gas use is expected, and is discussed in the [FAQs](https://github.com/sc-forks/solidity-coverage/blob/master/docs/faq.md#running-out-of-gas)

For everyone else:

* `testrpc-sc` should now be on par feature wise with `ganache-cli@6.4.1` (cc: @vdrg who was asking after this). It is likely there are some small technical differences between them, beyond the deliberate ones, but I would be surprised if they manifested in any significant way. If they do, please let me know. This is because I based the custom `ethereumjs-vm-sc` off of the most recent `ethereumjs-vm`, which isn't necessarily used in `ganache-cli`.
* If you are using a custom script to start `testrpc-sc`, you now need to add `--allowUnlimitedContractSize` to the arguments passed to `testrpc-sc`. You do not need to do this if you are using the `testrpcOptions` parameter in `.solcover.js`.

**Comment by JirkaChadima** [comment]
You, sir, are awesome! Outstanding work and it's now running as expected in our repo. Thank you very much!

File: lib/preprocessor.js (hunk 2 above)
Comment (area) [reply lib/preprocessor.js]: Now done, and though I didn't end up using your code, thank you for drawing attention to the issue!

**Comment by obernardovieira** [comment]
Hello @area 
It's been a while since I said I would get back with the problem I found, sorry for that.
Now the code is open-source (you can see here) and I'm having a new problem which is already reported *ParserError: extraneous input '{' expecting ')'*
This problem only happens with beta.5. In previous versions, the error is `There was a problem instrumenting ./coverageEnv/contracts/FixidityLib.sol: TypeError: Cannot read property 'stop' of null`
Anything I can help with? I'm available to fix it.

Thanks

**Comment by fernandomg** [comment]
Hello @area, thank you very much for the effort put into this project.

I'm facing an issue with this contract: https://github.com/GoodDollar/GoodContracts/blob/master/contracts/BancorFormula.sol

**The error:**
```
test-solidity-coverage on  master [✘] is 📦 v1.0.0 via ⬢ v10.15.3 
➜ npm run coverage

> test-solidity-coverage@1.0.0 coverage /.../test-solidity-coverage
> solidity-coverage

Generating coverage environment
Running: truffle compile  
(this can take a few seconds)...

Compiling your contracts...
===========================
> Compiling ./contracts/BancorFormula.sol
> Compiling ./contracts/ExpArray.sol
> Artifacts written to /.../test-solidity-coverage/coverageEnv/build/contracts
> Compiled successfully using:
   - solc: 0.5.3+commit.10d17f24.Emscripten.clang

Instrumenting  ./coverageEnv/contracts/BancorFormula.sol
Cleaning up...
There was a problem instrumenting ./coverageEnv/contracts/BancorFormula.sol: ParserError: missing ';' at '}' (133:36)
Exiting without generating coverage...

```

**To reproduce:**
```
git clone git@github.com:fernandomg/test-solidity-coverage
cd test-solidity-coverage
npm i && npm run coverage
```

Thanks!

**Comment by mrice32** [comment]
Hey @area, thanks for updating everything!

We're currently using the beta and we've been running into a compilation issue related to calling internal functions from a constructor (I think this used to be an issue in `solc`). The following contract fails to compile after being instrumented:
```
pragma solidity ^0.5.0;

pragma experimental ABIEncoderV2;

library TestLib {
    struct State {
        uint integer;
    }

    function internalFunction(State storage state) internal {
        state.integer = 5;
    }
}


contract Test {
    using TestLib for TestLib.State;
    TestLib.State state;

    constructor() public {
        state.internalFunction();
    }

    function getInteger() public returns (uint) {
        return state.integer;
    }
}
```
The error that the compiler gives is:
```
InternalCompilerError: Assembly exception for bytecode
```

We've noticed that moving the internal call to a `public` method seems to avoid the issue. The following contract compiles fine after being instrumented:
```
pragma solidity ^0.5.0;

pragma experimental ABIEncoderV2;

library TestLib {
    struct State {
        uint integer;
    }

    function internalFunction(State storage state) internal {
        state.integer = 5;
    }
}


contract Test {
    using TestLib for TestLib.State;
    TestLib.State state;

    constructor() public {
        callInternalFunction();
    }

    function getInteger() public returns (uint) {
        return state.integer;
    }

    function callInternalFunction() public {
        state.internalFunction();
    }
}
```

Two other things to note:
- We have also noticed that this only happens when ABIEncoderV2 is enabled.
- These contracts all compile (without instrumentation) with the most recent version of `solc`.

**Comment by daniel-iobuilders** [comment]
Hey @area,
I'm using the lates beta, but seem to have a bit of a strange behavior. The code coverage report shows a function with visibility internal as not covered, even though is it called several times during the tests.

The respective code can be found here: https://github.com/IoBuilders/holdable-token/tree/5a7bfac9bd55ffe91724fbeb06da4cfab5818255

After executing `npm install`, `npm run coverage` can be excuted to create the code coverage report. The report shows line 218 as not covered. This is not correct, as it is the only line of its function and the function is called several times. Not sure if it is important, but the function is a getter  for `block.timestamp`.

Thanks for any hints, how I could correct the code coverage report to be created correctly.

**Comment by cgewecke** [comment]
@daniel-iobuilders Is this the function?
```solidity
function _getBlockTimeStamp() internal view returns (uint256) {
        /* solium-disable-next-line security/no-block-members */
        return block.timestamp;
    }
```
It's possible the method is getting optimized out by solc. Is there a reason why you've encapsulated `block.timestamp` that way? Can the method be replaced by a direct reference to the variable?

**Comment by daniel-iobuilders** [comment]
> @daniel-iobuilders Is this the function?
> 
> ```solidity
> function _getBlockTimeStamp() internal view returns (uint256) {
>         /* solium-disable-next-line security/no-block-members */
>         return block.timestamp;
>     }
> ```
> 
> It's possible the method is getting optimized out by solc. Is there a reason why you've encapsulated `block.timestamp` that way? Can the method be replaced by a direct reference to the variable?

The encapsulation is done to overwrite it in the mock implementation of the contract and use it in the unit tests to manipulate the returned block timestamp. The smart contract has a concept of expiration and in order to test if the expiration logic I need to manipulate the returned timestamp.

I've deactivated the optimizer in truffle-config.js, but get the same result. I will do some more tests to check if it could be related to the compiler.

**Comment by cgewecke** [comment]
@daniel-iobuilders, re:
> use it in the unit tests to manipulate the returned block timestamp. The smart contract has a concept of expiration and in order to test if the expiration logic I need to manipulate the returned timestamp.

There is an alternative way of doing this with ganache which has extension RPC endpoints that let you increase the block time. OpenZeppelin publishes a [library of test helpers](https://github.com/OpenZeppelin/openzeppelin-test-helpers#time) you might find helpful for this purpose.

**Comment by yxliang01** [comment]
Is there a way to keep both the support of Solidity 0.5 and 0.4? While the newer version of Solidity is mandatory, would be great if the old stable version 0.4 can be retained until 0.6 is out.

**Comment by cgewecke** [comment]
@yxliang01 

> Is there a way to keep both the support of Solidity 0.5 and 0.4? While the newer version of Solidity is mandatory, would be great if the old stable version 0.4 can be retained until 0.6 is out.

Does this mean you'd like a version of the stable which has a newer ganache client?

**Comment by yxliang01** [comment]
Not only that. @cgewecke The PR description says that the modification in this PR breaks support of Solidity 0.4 at all. This indicates that upon merging this PR, all subsequent improvements to `solidity-coverage` can't be applied to projects requires `solc`. What I proposed is that don't break Solidity 0.4 support until support for Solidity 0.6 is ready.

**Comment by cgewecke** [comment]
@yxliang01 I've just created a branch from the current master called `v5-stable` where dependency updates for a solidity v4 compatible version will be possible. But I don't think we can provide backwards compatibility for more substantive improvements there.  

This PR is getting merged today because it *is* the improvement that's necessary.

**Comment by fabianorodrigo** [comment]
@cgewecke , this merge you're doing will end up with a new tag release and is going to be update on npmjs.org? My doubt is: when you finish, I can just execute 'npm install solidity-coverage' to get the merged package?
Thanks

**Comment by cgewecke** [comment]
@fabianorodrigo Yes. It's not published to npm yet, but will ping this PR when it is.

**Comment by cgewecke** [comment]
@fabianorodrigo Yes. It's not published to npm yet, but will ping this PR when it is.

**Comment by fabianorodrigo** [comment]
> Based on the [leapdao](https://github.com/leapdao/solidity-coverage#master) fork, with a nod to #311 as well. Also trying to get dependencies up-to-date.
> 
> This branch is currently published under the beta tag; you can try it out by `npm install --save --dev solidity-coverage@beta`. Bug reports would be very welcome before I release properly. You must be using solidity 0.5.1 or higher (0.5.0 will not work).

I received the following message when I tried to run solidity-coverage@beta on the project http://github.com/leapdao/solEVM-enforcer 

```
> Artifacts written to /home/fabianorodrigo/Projetos/ProjetosTerceiros/github/Solidity/20190626_GitResearchRepos/leapdao/solEVM-enforcer/coverageEnv/build/contracts
> Compiled successfully using:
   - solc: 0.5.2+commit.1df8f40c.Emscripten.clang

Instrumenting  ./coverageEnv/contracts/Enforcer.sol
Instrumenting  ./coverageEnv/contracts/EthereumRuntime.sol
Instrumenting  ./coverageEnv/contracts/EVMConstants.sol
Instrumenting  ./coverageEnv/contracts/EVMRuntime.sol
Cleaning up...
There was a problem instrumenting ./coverageEnv/contracts/EVMRuntime.sol: ParserError: no viable alternative at input 'els{' (97:17)
Exiting without generating coverage...

```

**Comment by cgewecke** [comment]
@fabianorodrigo That bug was fixed on master by #333. Am running the solEVM-enforcer tests rn on my box and they look ok (they instrumented fine and have passed a few hundred tests without problems). 

Hoping to publish 6.0 today or tomorrow, just waiting to see if the parser is going to get a new version tonight.

**Comment by cgewecke** [comment]
@fabianorodrigo 0.6.0 is published.

**Pull request merged**